src/abstract_charm.py

```diff
@@ -152,6 +166,115 @@ def wait_until_mysql_router_ready(self) -> None:
         else:
             logger.debug("MySQL Router is ready")
 
+    def _scope_obj(self, scope: secrets.Scopes) -> dict:
+        """Return corresponding data unit for app/unit."""
+        if scope == secrets.APP_SCOPE:
+            return self.app
+        if scope == secrets.UNIT_SCOPE:
+            return self.unit
+
+    def _peer_data(self, scope: secrets.Scopes) -> dict:
+        """Return corresponding databag for app/unit."""
+        if self.peers is None:
+            return {}
+        return self.peers.data[self._scope_obj(scope)]
+
+    def _get_secret_from_juju(self, scope: secrets.Scopes, key: str) -> typing.Optional[str]:
+        """Retrieve and return the secret from the juju secret storage."""
+        label = secrets.generate_secret_label(self, scope)
+        secret = self._secrets.get(label)
+
+        if not secret:
+            logger.debug("Getting a secret when secret is not added in juju")
+            return
+
+        value = secret.get_content().get(key)
+        logger.debug(f"Retrieved secret {key} for unit from juju")
+        return value
+
+    def _get_secret_from_databag(self, scope: secrets.Scopes, key: str) -> typing.Optional[str]:
+        """Retrieve and return the secret from the peer relation databag."""
+        return self._peer_data(scope).get(key)
+
+    def get_secret(self, scope: secrets.Scopes, key: str) -> typing.Optional[str]:
+        """Get secret from the secret storage.
+
+        Retrieve secret from juju secrets backend if secret exists there.
+        Else retrieve from peer databag. This is to account for cases where
+        secrets are stored in peer databag but the charm is then refreshed to
+        a newer revision.
+        """
+        if scope not in [secrets.APP_SCOPE, secrets.UNIT_SCOPE]:
+            raise MySQLRouterSecretsError(f"Invalid secret scope: {scope}")
+
+        if ops.jujuversion.JujuVersion.from_environ().has_secrets:
+            secret = self._get_secret_from_juju(scope, key)
+            if secret:
+                return secret
+
+        return self._get_secret_from_databag(scope, key)
```

Comment:
@juditnovak has plans to move all this code into a charm lib, but it is not there yet. i would be loath to make changes to this as this is standard across all charms (and would cause more pains down the road when we want to move to judit's charm lib)

Reply:
im unsure of the state of secrets and its libs, but i would like to iterate on this if we need to break it to a different lib. this PR is already too bloated